Lib/traceback.py

```diff
@@ -1081,50 +1081,33 @@ def __init__(self, exc_type, exc_value, exc_traceback, *, limit=None,
 
         self._is_syntax_error = False
         self._have_exc_type = exc_type is not None
-        if exc_type is not None:
+
+        if self._have_exc_type:
             self.exc_type_qualname = exc_type.__qualname__
             self.exc_type_module = exc_type.__module__
+            if issubclass(exc_type, SyntaxError):
+                # Handle SyntaxError's specially
+                self.filename = exc_value.filename
+                lno = exc_value.lineno
+                self.lineno = str(lno) if lno is not None else None
+                end_lno = exc_value.end_lineno
+                self.end_lineno = str(end_lno) if end_lno is not None else None
+                self.text = exc_value.text
+                self.offset = exc_value.offset
+                self.end_offset = exc_value.end_offset
+                self.msg = exc_value.msg
+                self._is_syntax_error = True
+                self._exc_metadata = getattr(exc_value, "_metadata", None)
+            elif suggestion := _compute_suggestion_message(exc_type, exc_value, exc_traceback):
+                if self._str.endswith(('.', '?', '!', '...')):
```

Comment:
Ellipsis is covered by the full stop.
```suggestion
                if self._str.endswith(('.', '?', '!')):
```